app_dart/lib/src/service/github_service.dart

```diff
@@ -15,27 +15,40 @@ class GithubService {
   final GitHub github;
 
   /// Lists the commits of the provided repository [slug] and [branch].
-  Future<List<RepositoryCommit>> listCommits(
-      RepositorySlug slug, String branch, int hours, Config config) async {
-    final Duration time = Duration(hours: hours);
+  Future<List<RepositoryCommit>> listCommits(RepositorySlug slug, String branch,
+      int lastCommitTimestampMills, Config config) async {
     ArgumentError.checkNotNull(slug);
     final PaginationHelper paginationHelper = PaginationHelper(github);
 
+    /// Return only one page when this is a new branch. Otherwise it will
+    /// return all commits prior to this release branch commit, leading to
+    /// heavy workload.
+    int pages;
+    if (lastCommitTimestampMills == 0) {
+      pages = 1;
+    }
+
     List<Map<String, dynamic>> commits = <Map<String, dynamic>>[];
+
+    /// [lastCommitTimestamp+1] excludes last commit itself.
     await for (Response response in paginationHelper.fetchStreamed(
       'GET',
       '/repos/${slug.fullName}/commits',
       params: <String, dynamic>{
         'sha': branch,
-        'since': DateTime.now().toUtc().subtract(time).toIso8601String()
+        'since':
+            DateTime.fromMillisecondsSinceEpoch(lastCommitTimestampMills + 1)
+                .toUtc()
+                .toIso8601String(),
       },
+      pages: pages,
     )) {
       commits.addAll((json.decode(response.body) as List<dynamic>)
           .cast<Map<String, dynamic>>());
     }
 
     /// Take the latest single commit for a new release branch.
-    if (hours == config.newBranchHours) {
+    if (lastCommitTimestampMills == 0) {
       commits = commits.take(1).toList();
```

Comment:
There could be more than one commits in a new release branch.

Reply:
Only the most recent commit is needed: https://github.com/flutter/cocoon/pull/734#discussion_r405195385

Reply:
That's a heuristic and I understand that works based on the assumption that Cocoon processes every new commit. However if Cocoon is down for a while (the reason of this PR), there can be multiple unprocessed commits in the new branch.

Also it'd better to explain every heuristics in comment, so that future readers won't have trouble in understanding the assumption and reasoning.

Reply:
Ah, you meant new commits submitted to the branch, that has branched out from main. This case does exist, but it usually occurs after couple of days since branching out.

It is a rare case that our cocoon service keeps down for days, but this corner case still may happen. Considering this, I would prefer to have this as a TODO, since this is not an easy implementation and we are not in need of this urgently.

At the same time, have added doc to explain.

Reply:
TODO added.